fe/fe-core/src/main/java/com/starrocks/leader/Checkpoint.java

```diff
@@ -79,147 +83,135 @@ public Checkpoint(String name, Journal journal, String subDir, boolean belongToG
-                int port = Config.http_port;
+        // Step3: Delete old journals
+        // conditions: 1. new image created and no others node to push, this means there is only one FE in the cluster,
+        //                delete the old journals immediately.
+        //             2. needToPushCnt > 0 means there are other nodes in the cluster,
+        //                we must make sure all the other nodes have got the new image and then delete old journals.
+        if ((newImageCreated && needToPushCnt == 0)
+                || (needToPushCnt > 0 && nodesToPushImage.size() == 0)) {
+            deleteOldJournals(newImageVersion);
+        }
 
-                String url = "http://" + host + ":" + port + "/put?version=" + checkpointVersion
-                        + "&port=" + port + "&subdir=" + subDir;
-                LOG.info("Put image:{}", url);
+        // Step4: Delete old image files from local storage.
+        if (newImageCreated) {
+            MetaCleaner cleaner = new MetaCleaner(imageDir);
+            try {
+                cleaner.clean();
+            } catch (IOException e) {
+                LOG.error("Leader delete old image file fail.", e);
+            }
+        }
+    }
 
-                try {
-                    MetaHelper.getRemoteFile(url, PUT_TIMEOUT_SECOND * 1000, new NullOutputStream());
-                    successPushed++;
-                } catch (IOException e) {
-                    LOG.error("Exception when pushing image file. url = {}", url, e);
-                }
+    private void deleteOldJournals(long imageVersion) {
+        // To ensure that all nodes will not lose data,
+        // deleteVersion should be the minimum value of imageVersion and replayedJournalId.
+        long minReplayedJournalId = getMinReplayedJournalId();
+        long deleteVersion = Math.min(imageVersion, minReplayedJournalId);
+        journal.deleteJournals(deleteVersion + 1);
+        LOG.info("journals <= {} with prefix [{}] are deleted. image version {}, other nodes min version {}",
+                deleteVersion, journal.getPrefix(), imageVersion, minReplayedJournalId);
+
+    }
+
+    private void pushImage(long imageVersion) {
+        Iterator<String> iterator = nodesToPushImage.iterator();
+        int needToPushCnt = nodesToPushImage.size();
+        int successPushedCnt = 0;
+        while (iterator.hasNext()) {
+            String nodeName = iterator.next();
+
+            Frontend frontend = GlobalStateMgr.getServingState().getNodeMgr().getFeByName(nodeName);
+            if (frontend == null) {
+                iterator.remove();
+                continue;
             }
 
-            LOG.info("push image.{} from subdir [{}] to other nodes. totally {} nodes, push succeeded {} nodes",
-                    checkpointVersion, subDir, otherNodesCount, successPushed);
-        }
+            String url = "http://" + frontend.getHost() + ":" + Config.http_port + "/put?version=" + imageVersion
```

Comment:
always push to all other frontend nodes without checking whether it's necessary? it may work but not efficient